compiler/vm/vm.nim

```diff
@@ -107,145 +181,236 @@ proc stackTraceImpl(
+  of akInt:
+    reg = TFullReg(kind: rkInt, intVal: v.intVal)
+  of akFloat:
+    reg = TFullReg(kind: rkFloat, floatVal: v.floatVal)
+  of akPtr:
+    reg = TFullReg(kind: rkAddress, addrVal: makeHandle(v.ptrVal, v.typ.targetType.sizeInBytes), addrTyp: v.typ.targetType, isValid: true)
+  of akDiscriminator:
+    reg = TFullReg(kind: rkInt, intVal: readDiscriminant(v.atom, v.typ.sizeInBytes, v.typ.numBits))
+  of akPNode:
+    # TODO: ensure that only NimNodes go into registers (typedescs are akPNodes too)
+    reg = TFullReg(kind: rkNimNode, nimNode: v.nodeVal)
+  else:
+    var handle: TypedHandle
+    if v.kind in realAtomKinds:
+      handle = mm.allocAtomLocation(v.typ)
+      copyToLocation(dest = handle, src = getHandleToValue(v), mm)
+    else:
+      handle = makeTHandle(move v.obj, v.typ)
+    
+    reg = TFullReg(kind: rkValue, handle: handle)
+
+
+# TODO: needs a better name
+func putIntoReg(reg: var TFullReg, h: TypedHandle) =
+  let atom = addr checkedDeref(h.h, h.typ.sizeInBytes)
+  case h.typ.kind
+  of akInt:
+    reg = TFullReg(kind: rkInt, intVal: atom.intVal)
+  of akFloat:
+    reg = TFullReg(kind: rkFloat, floatVal: atom.floatVal)
+  of akPtr:
+    reg = TFullReg(kind: rkAddress, addrVal: makeHandle(atom.ptrVal, h.typ.targetType.sizeInBytes), addrTyp: h.typ.targetType, isValid: false)
+  of akDiscriminator:
+    reg = TFullReg(kind: rkInt, intVal: readDiscriminant(h))
+  of akPNode:
+    # TODO: ensure that only NimNodes go into registers (typedescs are akPNodes too)
+    reg = TFullReg(kind: rkNimNode, nimNode: atom.nodeVal)
+  else:
+    reg = TFullReg(kind: rkHandle, handle: h)
+
+proc reportException(c: var TCtx; tos: StackFrameIndex, raised: TypedHandle, excType: PType) =
   # REFACTOR VM implementation relies on the `stackTrace` calling return,
   # but in this proc we are retuning only from it's body, so calling
   # `reportException()` does not stop vm loops. This needs to be cleaned up
   # - invisible injection of the `return` to control flow of execution is
   # an absolute monkey-tier hack.
   stackTrace(
     c, tos, c.exceptionInstr,
-    reportAst(rsemVmUnhandledException, raised))
-
+    reportAst(rsemVmUnhandledException, unmarshalException(raised, excType)))
+
+
+proc reportAddressCheckFail(c: var TCtx, tos: StackFrameIndex, pc: int, reason: AddressCheckFailReason) =
+  # TODO: should probably use a dedicated report kind?
+  case reason
+  of acfrOutOfBounds:
+    stackTrace(
+      c, tos, pc,
+      reportStr(rsemVmErrInternal, "Accessing out-of-bounds location"))
+  of acfrTypeMismatch:
+    # TODO: more descriptive error message
+    stackTrace(
+      c, tos, pc,
+      reportStr(rsemVmErrInternal, "Mismatching type between pointer and location")
+    )
 
 when not defined(nimComputedGoto):
   {.pragma: computedGoto.}
 
-proc ensureKind(n: var TFullReg, k: TRegisterKind) {.inline.} =
+func updateRegisters(c: var TCtx) =
+  ## Invalidate all handles and addresses that reference now freed regions
```

Comment:
Removing this would mean that guest code is able to access host memory via pointers/handles to now freed memory regions. Consider:
```nim
var a = @[1]
let p = addr a[0]
reset(a)
p[] = 2
```
`updateRegister` invalidates `p` after the reset, which forces a re-validation at `p[]` which would fail. Instead of setting handles to nil (which violates their invariants, as they're required to always be valid), the register containing the now invalid handle should simply be reset to `rkNone`

An alternative would be to always validate handles prior to using them, but, as described in a comment above, that has a rather high cost.

My current plan is to lift the logic from `updateRegisters` into a dedicated instruction that would explicitly be inserted by `vmgen` at strategic positions.

As yet another alternative, pointer/handle validity checks could also be moved into a dedicated instruction which would be inserted by `vmgen` prior to pointer/handle usage; implicit validation/re-validation would be completely removed.

Reply:
> As yet another alternative, pointer/handle validity checks could also be moved into a dedicated instruction which would be inserted by `vmgen` prior to pointer/handle usage; implicit validation/re-validation would be completely removed.

I'd prefer something like this, eliminating checks based on statically computed pre-conditions rather than dynamically caching a state.